### PR TITLE
Renderer/HorizonRenderer: more elaborate artificial horizon

### PR DESCRIPTION
**Pull request by qubolino**
more elaborate artificial horizon with standard colorblind-friendly colors, shape and marks

Brief summary of the changes
----------------------------

Related issues and discussions
------------------------------

### CODE REVIEW

**Comment by MaxKellermann** [comment]
You requested a review, but your code was already reviewed in https://github.com/XCSoar/XCSoar/pull/1261 and apparently you didn't obey the review. I wish you hadn't closed https://github.com/XCSoar/XCSoar/pull/1261 because now we're wasting time on doing the same again.

**Comment by qubolino** [comment]
i did spend quite some time complying to what lordfolken advised. 

if that still needs work to meet your standards, i suggest you do it yourself, or ditch my work that i happily tried to share with the community

File: src/Look/HorizonLook.cpp [hunk 1]
```diff
@@ -7,11 +7,17 @@
 void
 HorizonLook::Initialise()
 {
-  aircraft_pen.Create(Layout::Scale(2), COLOR_BLACK);
+  aircraft_pen.Create(Layout::Scale(1), Color(0xf2, 0xf2, 0x19));
+  aircraft_brush.Create(Color(0xf2, 0xf2, 0x19));
```
Comment (lordfolken) [inline src/Look/HorizonLook.cpp:11]: Plsease make a definition for Color(0xf2, 0xf2, 0x19)  in HorizionLook.hpp, and reuse the definition here.

**Comment by MaxKellermann** [comment]
> i did spend quite some time complying to what lordfolken advised.

Just look at the **very first** review comment: https://github.com/XCSoar/XCSoar/pull/1261#discussion_r1253499298

Then look at this piece in this PR you just submitted:
```
   // sky_pen.Create(Layout::Scale(1), LightColor(sky_color));
```
The is the very line @lordfolken criticized, and it's still there. You didn't reply to his comment and you did not fix this.

I stopped reading your PR after seeing this, but I should have stopped reading after seeing that your commit message doesn't even explain what your proposed change is about.

File: src/Look/HorizonLook.cpp [hunk 2]
```diff
@@ -7,11 +7,17 @@
...
+  mark_pen.Create(Layout::Scale(1), COLOR_WHITE);
+  mark_brush.Create(COLOR_WHITE);
+  horizon_pen.Create(Layout::Scale(1), COLOR_WHITE);
 
   sky_brush.Create(sky_color);
-  sky_pen.Create(Layout::Scale(1), DarkColor(sky_color));
+  // sky_pen.Create(Layout::Scale(1), LightColor(sky_color));
```
Comment (lordfolken) [inline src/Look/HorizonLook.cpp:17]: Please remove commented line

File: src/Look/HorizonLook.cpp [hunk 3]
```diff
@@ -7,11 +7,17 @@
...
+  sky_pen.Create(Layout::Scale(1), sky_color);
 
   terrain_brush.Create(terrain_color);
-  terrain_pen.Create(Layout::Scale(1), DarkColor(terrain_color));
+  // terrain_pen.Create(Layout::Scale(1), LightColor(terrain_color));
```
Comment (lordfolken) [inline src/Look/HorizonLook.cpp:21]: Please remove commented line

File: src/Renderer/HorizonRenderer.cpp [hunk 4]
```diff
@@ -11,6 +11,86 @@
 
 #include <algorithm>
 
+#include <iostream>
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.cpp:14]: Do we need this include? I haven't seen a function that uses it, but haven't tried to remove it yet.

File: src/Renderer/HorizonRenderer.hpp [hunk 5]
```diff
@@ -8,9 +8,21 @@ struct HorizonLook;
...
             const AttitudeState &attitude);
+
+int lines_intersect(PixelPoint p1, PixelPoint p2, 
+                    PixelPoint p3, PixelPoint p4, 
+                    PixelPoint &intersect);
+
+void rotate(PixelPoint point, PixelPoint center, 
+            Angle a, PixelPoint &rotated);
+
+
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.hpp:27]: Surplus line. Please remove.

File: src/Renderer/HorizonRenderer.cpp [hunk 6]
```diff
@@ -41,36 +121,190 @@ HorizonRenderer::Draw(Canvas &canvas, const PixelRect &rc,
...
-  auto sphi = Angle::HalfCircle() - Angle::Degrees(bank_degrees);
-  auto alpha1 = sphi - alpha;
-  auto alpha2 = sphi + alpha;
-
-  // draw sky part
-  if (cosine_ratio > -1 ) { // when less than -1 then the sky is not visible
-    canvas.Select(look.sky_pen);
-    canvas.Select(look.sky_brush);
-    canvas.DrawSegment(center, radius, alpha2, alpha1, true);
+
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.cpp:124]: Surplus line.

**Comment by MaxKellermann** [comment]
Oh no, now you suggest merging a merge commit. Everything is wrong with this PR.

I suggest you should continue working on https://github.com/XCSoar/XCSoar/pull/1261 and not this one, because over there, you see unresolved review comments.

Your last post sounded like you don't want to do that. That's okay for me, you can do whatever you want, and you don't need to listen to us. But it might cause us not listening to you either.

**Review by lordfolken — CHANGES_REQUESTED** [review]
Thank you for sticking with it. A few minor points in my current review. 
Next i'll try to compile and run your code.

**Comment by lordfolken** [comment]
Git history:
we like rebased code on current master. Merges aren't optimal. 
do a 
git pull origin master --rebase on your branch, commit your code and do a force push here.

**Comment by lordfolken** [comment]
Max isn't very tolerant of when things aren't done in the way he expects. 
He is however fair when the code is well written (and i thnk the code is clear and good) and will merge it. 
I'll try to help get things into shape with you.
So just stick with it and don't get discouraged.

**Comment by lordfolken** [comment]
@MaxKellermann i'll reopen, when i think it looks presentable.

**Comment by lordfolken** [comment]
@iltis42 and @XCNav can you have a look at this? i think your customers would benefit mostly.

**Comment by MaxKellermann** [comment]
> Max isn't very tolerant of when things aren't done in the way he expects.

First of all, I expect people to handle the existing reviews before asking for another review. @qubolino ignored the previous review, while pretending he "complied" (which I proved wrong). Then he explained he didn't want to meet our standards and asked me to do it for him. I certainly won't do that.

And this isn't event arguing whether commented line would be acceptable. At this point, this isn't about the code at all. It's about @qubolino's behavior towards code reviews.

**Comment by qubolino** [comment]
dear max. i suggest you try be more tolerant on others skills. 

i respect your work, and you probably are a very skilled programmer. 

i opened a new pr b/c i've been asked to put everything in one commit. i don't know how to do this working with an existing pr. i actually don't know how to work with git as you already know. in fact i don't know how to do many things. 

one thing i know though is how to propose the perhaps only mediocre result of my work to the community -- despite the fact i benefited from a surely high standard result of its work. this pr was initially more a pull proposition than a request. i initially didn't envision to do anything more on it. then i thought it was maybe not too hard to comply and could make an effort and i did.

**Comment by qubolino** [comment]
> Max isn't very tolerant of when things aren't done in the way he expects. 
> 
> He is however fair when the code is well written (and i thnk the code is clear and good) and will merge it. 
> 
> I'll try to help get things into shape with you.
> 
> So just stick with it and don't get discouraged. 

he isn't very tolerant of probably many more things than this. 
he imvho has a gift though. he is very good at discouraging people to contribute to something he doesn't own. the community does. fine by me. three times i interacted with him, three times i seemed to be completely wrong and felt my work was treated disrespectfully.

**Comment by MaxKellermann** [comment]
@qubolino you're trying to reframe your failure to read review comments carefully, making it sound like it's about me. Please don't make this personal. This doesn't help, you're only wasting more of everybody's time with a pointless off-topic discussion that leads nowhere.

(You have annoyed me enough that I'm going to unsubscribe from this PR and I'm not going to spend more time on reviewing it. Bye.)

**Comment by qubolino** [comment]
i find your behaviour very childish and non-constructive. but that's only my opinion.

speaking about being constructive, @lordfolken thanks a lot for (twice already) trying to encourage me to carry on. i'll be happy to work with you.

**Comment by qubolino** [comment]
so #1 seems to make the pr square. which should i use? how do i make it acceptable for you (not even talking about the code content yet)?
only one commit, no merge? thanks for helping and for applying your pedagogical skills.

**Comment by qubolino** [comment]
@lordfolken thanks for your comments, i addressed them, i think.

looking fwd to your constructive critics

**Comment by DanD222** [comment]
@qubolino 
If you look here https://github.com/XCSoar/XCSoar/pull/1266/commits, you'll see two new commits called "addressing" - you want to avoid that (as well as the "_Merge branch 'XCSoar:master' into master_" commit). If the PR was merged in this state, then XCSoar's commit history would include a whole bunch of "addressing" commits, as well as the wrong commits they fix.


What you might want to do instead is squash the "addressing" commit with the commit where the code that you're addressing is - it then becomes one commit with the right/addressed code.

After that, you Force Push from your local machine to your online Git repo - this rewrites the commits in the Pull Request, gets rid of the extraneous commits, and provides a nice commit set for review (providing the commit history is correct on your machine in the first place).

Regarding the "_Merge branch 'XCSoar:master' into master_" commit, you want to Rebase your branch on origin/master instead - this brings your branch up to date, but doesn't include the unnecessary merge commit.

(Not sure what your workflow is and sorry if I'm stating the obvious, but you can't do any of this through the Github web UI - this needs to be done on your local machine through an IDE/Git GUI/Git commands)

**Comment by qubolino** [comment]
i squashed everything in one commit

**Comment by lordfolken** [comment]
Finally got around to testing it. So the main window looks good to me and seems to behave properly. The Attitude Infobox, has some errors. Its not filling the whole square and the lines are not adapting to the smaller resolution. 
Your commit is missing a new line at the end.

**Comment by lordfolken** [comment]
Finally got around to testing it. So the main window looks good to me and seems to behave properly. The Attitude Infobox, has some errors. Its not filling the whole square and the lines are not adapting to the smaller resolution. 
Your commit is missing a new line at the end.

**Comment by qubolino** [comment]
> the yellow arrow should as well move with the bank as long the yaw is not changing
in old mechanical (but also traditionally) the yellow is not moving relative to the plane (nor is anything due to the yaw)
> The small infobox is overloaded
agreed and addressed. i kept only ground sky and plane in yellow

**Comment by qubolino** [comment]
> the yellow arrow should as well move with the bank as long the yaw is not changing

in old mechanical (but also traditionally) the yellow is not moving relative to the plane (nor is anything due to the yaw)

> The small infobox is overloaded

agreed and addressed. i kept only ground sky and plane in yellow

**Comment by XCNav** [comment]
Screenshot_20231012_084527_Photos
Here you can see how yaw is handled at modern EFIS Aircraft Avionics. The yaw indicator is the little white bar below the white bank angle arrow. They both move basically in one unit. As the Aircraft slips the white bar is diverting from its position to the left or to the right. Maximum deflection is one bar width. Is the nose yawing to the left, the bar is diverting to the right and vice versa.

**Comment by qubolino** [comment]
> Here you can see how yaw is handled at modern EFIS Aircraft Avionics

what you describe isn't yaw https://en.wikipedia.org/wiki/Yaw_(rotation). 
it's a slip indicator.
and there are different conventions for different efis.

my implementation of the horizon only takes into account roll and pitch. 
and i chose the convention of yellow isn't moving relative to aircraft.
https://en.wikipedia.org/wiki/Attitude_indicator

File: src/Renderer/HorizonRenderer.cpp [hunk 7]
```diff
@@ -278,19 +281,37 @@ HorizonRenderer::Draw(Canvas &canvas, const PixelRect &rc,
...
       rotate(m1, center, Angle::Degrees(- bank_degrees), p1);
       rotate(m2, center, Angle::Degrees(- bank_degrees), p2);
       canvas.DrawLine(p1,p2);
+
+      canvas.Select(look.mark_font);
+      canvas.SetBackgroundTransparent();
+      canvas.SetTextColor( COLOR_WHITE);
+      char buffer[4];
+      sprintf(buffer, "%+3d", k);
+      canvas.DrawText(p2, buffer);
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.cpp]: Buffer overflow when actually using the formatting at its length. Don't forgett the Null Terminator.

File: src/Renderer/HorizonRenderer.cpp [hunk 8]
```diff
@@ -278,19 +281,37 @@ HorizonRenderer::Draw(Canvas &canvas, const PixelRect &rc,
...
+    for (int k=pitch_10-25; k<=pitch_10+25; k+=10){
+      m1 = {-radius/10, int(h/2.f*(pitch_degrees-k)/MAX_HEIGHT_DEGREES)};
+      m2 = {+radius/10, int(h/2.f*(pitch_degrees-k)/MAX_HEIGHT_DEGREES)};
       rotate(m1, center, Angle::Degrees(- bank_degrees), p1);
       rotate(m2, center, Angle::Degrees(- bank_degrees), p2);
       canvas.DrawLine(p1,p2);
+
+      canvas.Select(look.mark_font);
+      canvas.SetBackgroundTransparent();
+      canvas.SetTextColor( COLOR_WHITE);
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.cpp]: Space uneccessary

File: src/Renderer/HorizonRenderer.cpp [hunk 9]
```diff
@@ -278,19 +281,37 @@ HorizonRenderer::Draw(Canvas &canvas, const PixelRect &rc,
...
       rotate(m1, center, Angle::Degrees(- bank_degrees), p1);
       rotate(m2, center, Angle::Degrees(- bank_degrees), p2);
       canvas.DrawLine(p1,p2);
+
+
+      canvas.Select(look.mark_font);
+      canvas.SetBackgroundTransparent();
+      canvas.SetTextColor( COLOR_WHITE);
+      char buffer[4];
+      sprintf(buffer, "%+3d", k);
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.cpp:334]: Potential buffer overflow

File: src/Renderer/HorizonRenderer.cpp [hunk 10]
```diff
@@ -23,6 +23,8 @@
 
 #define MAX_HEIGHT_DEGREES 40.f
 
+char buffer[3];
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.cpp:28]: Unused?

File: src/Renderer/HorizonRenderer.cpp [hunk 11]
```diff
@@ -142,21 +144,35 @@ HorizonRenderer::Draw(Canvas &canvas, const PixelRect &rc,
   BulkPixelPoint corners[] = {k21, k11, k12, k22, k21};
   PixelPoint intersect_left = {0, 0};
   PixelPoint intersect_right = {0, 0};
-  for (i=0; i < 4; i++){
+  for (i=0; i <= 4; i++){
     int res = lines_intersect(p1, p3, corners[i], 
                               corners[i+1], intersect_left);
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.cpp:149]: Out of bounds for corners[5]

File: src/Renderer/HorizonRenderer.cpp [hunk 12]
```diff
@@ -9,6 +9,7 @@
 #include "NMEA/Attitude.hpp"
 #include "Math/Constants.hpp"
 #include "Math/Util.hpp"
+#include "LogFile.hpp"
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.cpp:12]: Debugging?

File: src/Look/HorizonLook.hpp [hunk 13]
```diff
@@ -7,15 +7,23 @@
...
   Brush sky_brush;
   Pen sky_pen;
 
-  static constexpr Color terrain_color{0x80, 0x45, 0x15};
+  static constexpr Color terrain_color{0x78, 0x51, 0x2a};
   Brush terrain_brush;
   Pen terrain_pen;
 
+  static constexpr Color aircraft_color{0xf2, 0xf2, 0x19};
+  Pen aircraft_pen;
```
Comment (lordfolken) [inline src/Look/HorizonLook.hpp:21]: Please keep color definitions in Look/HorizonLook.hpp

File: src/Renderer/HorizonRenderer.hpp [hunk 14]
```diff
@@ -8,9 +8,20 @@ struct HorizonLook;
...
-}
+
+int lines_intersect(PixelPoint p1, PixelPoint p2, 
+                    PixelPoint p3, PixelPoint p4, 
+                    PixelPoint &intersect);
+
+void rotate(PixelPoint point, PixelPoint center, 
+            Angle a, PixelPoint &rotated);
+
+}
```
Comment (lordfolken) [inline src/Renderer/HorizonRenderer.hpp:27]: Missing newline

**Review by lordfolken — CHANGES_REQUESTED** [review]

**Comment by lordfolken** [comment]
@qubolino if you are still around, can you have a look?

**Comment by qubolino** [comment]
my focus was elsewhere @lordfolken. thanks a lot for merging!
will try work better with you guys in the future

**Comment by lordfolken** [comment]
@qubolino i'm sure there is bugs and optimizations that are needed in there. Also further functionality such as  a speed/altitude tape etc. so if you would continue to maintain and help, i'd be very happy.

**Pull request merged**